src/grpc/pp_ics_gateway_location_worker.erl

```diff
@@ -0,0 +1,231 @@
+-export([
+    start_link/1,
+    init_ets/0,
+    get/1
+]).
+
+%% ------------------------------------------------------------------
+%% gen_server Function Exports
+%% ------------------------------------------------------------------
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-define(ICS_CHANNEL, ics_channel).
+-define(SERVER, ?MODULE).
+-define(ETS, pp_ics_gateway_location_worker_ets).
+-define(INIT, init).
+-ifdef(TEST).
+-define(BACKOFF_MIN, 100).
+-else.
+-define(BACKOFF_MIN, timer:seconds(10)).
+-endif.
+-define(BACKOFF_MAX, timer:minutes(5)).
+
+-record(state, {
+    pubkey_bin :: libp2p_crypto:pubkey_bin(),
+    sig_fun :: function(),
+    transport :: http | https,
+    host :: string(),
+    port :: non_neg_integer(),
+    conn_backoff :: backoff:backoff()
+}).
+
+-record(location, {
+    gateway :: libp2p_crypto:pubkey_bin(),
+    timestamp :: non_neg_integer(),
+    h3_index :: h3:index()
+}).
+
+-type state() :: #state{}.
+
+%% ------------------------------------------------------------------
+%% API Function Definitions
+%% ------------------------------------------------------------------
+
+start_link(Args) ->
+    case start_link_args(Args) of
+        ignore ->
+            lager:warning("~s ignored ~p", [?MODULE, Args]),
+            ignore;
+        Map ->
+            gen_server:start_link({local, ?SERVER}, ?SERVER, Map, [])
+    end.
+
+-spec init_ets() -> ok.
+init_ets() ->
+    ?ETS = ets:new(?ETS, [
+        public,
+        named_table,
+        set,
+        {read_concurrency, true},
+        {keypos, #location.gateway}
+    ]),
+    ok.
+
+-spec get(libp2p_crypto:pubkey_bin()) -> {ok, h3:index()} | {error, any()}.
+get(PubKeyBin) ->
+    case lookup(PubKeyBin) of
+        {error, _Reason} ->
+            gen_server:call(?SERVER, {get, PubKeyBin});
+        {ok, _} = OK ->
+            OK
+    end.
+
+%% ------------------------------------------------------------------
+%% gen_server Function Definitions
+%% ------------------------------------------------------------------
+init(
+    #{
+        pubkey_bin := PubKeyBin,
+        sig_fun := SigFun,
+        transport := Transport,
+        host := Host,
+        port := Port
+    } = Args
+) ->
+    lager:info("~p init with ~p", [?SERVER, Args]),
+    Backoff = backoff:type(backoff:init(?BACKOFF_MIN, ?BACKOFF_MAX), normal),
+    %% self() ! ?INIT,
+    {ok, #state{
+        pubkey_bin = PubKeyBin,
+        sig_fun = SigFun,
+        transport = Transport,
+        host = Host,
+        port = Port,
+        conn_backoff = Backoff
+    }}.
+
+handle_call({get, PubKeyBin}, _From, #state{conn_backoff = Backoff0} = State) ->
+    HotspotName = pp_utils:animal_name(PubKeyBin),
+    case get_gateway_location(PubKeyBin, State) of
+        {error, Reason, true} ->
+            {Delay, Backoff1} = backoff:fail(Backoff0),
+            _ = erlang:send_after(Delay, self(), ?INIT),
```

Comment:
Probably should cleanup the all init thing if we are just using the `.config` to start the channel